### ISSUE TITLE
can't converge？

### ISSUE THREAD

**Issue by MenglaiWang** [shiyemin/ResNet-Generator-for-caffe#1]
@shiyemin  I have use you generator to create a ResNet prototxt, but after 54000 iterations , I have only get the acc/top-1 = 0.0012 and acc/top-5 = 0.0066, the loss is around 7.
![image](https://cloud.githubusercontent.com/assets/11674319/12006562/839cdd12-ac16-11e5-9289-583dbd045c70.png)

Do you have any idea where I got wrong?

**Comment by shiyemin** [comment]
@MenglaiWang As you can see, this model is hard to train. In my practice, I successfully train ResNet after adding another 2 loss layer as what GoogLenet did. Hope this can help you.

**Comment by MenglaiWang** [comment]
@shiyemin Thanks a lot ! Also, in the prototxt, training batchsize is 8, it can't tolerate a big learning rate. So, when I changed lr into 0.01,it starts to converge. To be honest, I didn't get the competitive result as in their paper. After 600,000 iterations, I only got acc/top-5 = 0.782196 and acc/top-1 = 0.541999 .
How about you?

**Comment by shiyemin** [comment]
Yeah, you are right. That's why i choose to train resnet with mxnet. I train ResNet with a batch size of 256 using 8 Tesla K80. With big batch size (256), more loss layer (3 loss layer), it is able to converge even using 0.1 lr. But it will take 2-3 week to finish training. The accuracy after 7 epochs is 23.05%. Maybe i will publish my model if everything works well (if no body publish ResNet model before me).

**Comment by shiyemin** [comment]
@MenglaiWang By the way, what i'm training is the 152-layer model. It will take much more time than the shallower model.

**Comment by MenglaiWang** [comment]
@shiyemin That would be very helpful.

**Comment by smichalowski** [comment]
@shiyemin could you please share your prototxt with added 2 loss layers ? Im not sure where to add them. Thanks

**Comment by shiyemin** [comment]
@smichalowski I add loss layers after the last two layers where  the size of feature maps changed (right before projection shortcut). I remove these additional loss layer when the loss stop decreasing. Then the network could be trained in a standard way.

Because i'm training this network with mxnet now, i don't have a modified version of Caffe prototxt.

**Comment by smichalowski** [comment]
@shiyemin Did you have similiar results ? My lr is 0.01. 

I0124 23:40:57.258854 12332 solver.cpp:341] Iteration 318000, Testing net (#0)
I0124 23:55:12.107619 12332 solver.cpp:409]     Test net output #0: acc/top-1 = 0.212104
I0124 23:55:12.107806 12332 solver.cpp:409]     Test net output #1: acc/top-5 = 0.4304
I0124 23:55:12.107831 12332 solver.cpp:409]     Test net output #2: loss = 4.37948 (* 1 = 4.37948 loss)

![zrzut ekranu 2015-11-04 o 18 46 01](https://cloud.githubusercontent.com/assets/3269993/12539638/fa4f7256-c2f7-11e5-9bfa-8668a5a2d109.png)

**Comment by smichalowski** [comment]
@shiyemin Did you have similiar results ? My lr is 0.01.

I0124 23:40:57.258854 12332 solver.cpp:341] Iteration 318000, Testing net (#0)
I0124 23:55:12.107619 12332 solver.cpp:409] Test net output #0: acc/top-1 = 0.212104
I0124 23:55:12.107806 12332 solver.cpp:409] Test net output #1: acc/top-5 = 0.4304
I0124 23:55:12.107831 12332 solver.cpp:409] Test net output #2: loss = 4.37948 (* 1 = 4.37948 loss)

![your_chart_name](https://cloud.githubusercontent.com/assets/3269993/12539661/4a3ff740-c2f8-11e5-83e5-8238d006cc32.png)

**Comment by shiyemin** [comment]
@smichalowski Yeah, I believe your accuracy is good but loss is not stable. I'm not sure why this happening. Maybe your batchsize is a little too small?

**Comment by smichalowski** [comment]
@shiyemin my train batch size is 8. val batch size is 5. After next few hours things get weird:

![your_chart_name](https://cloud.githubusercontent.com/assets/3269993/12551421/95da5548-c36a-11e5-9a0c-f2e2b47e298c.png)

**Comment by shiyemin** [comment]
@smichalowski My advice is to use mxnet or more GPUs and this will allow you to use a bigger batchsize. We have tested that a batchsize of 32 will get this model converge with 0.01 learning rate.

**Comment by Fchaubard** [comment]
Is there any chance you can provide some pretrained caffemodel's? I would love to be able to deploy a resnet model without having to download and untar imagenet. Thanks!

**Comment by smichalowski** [comment]
@Fchaubard as @shiyemin said, resnet is very hard to train without powerful hardware. You can find already trained models [here](https://onedrive.live.com/?authkey=%21AAFW2-FVoxeVRck&id=4006CBB8476FF777%2117887&cid=4006CBB8476FF777)

**Comment by etienne87** [comment]
hello there, why didn't you use net_spec from caffe?
`top = L.Convolution(input,...)`
also, it seems we need a Scale Layer after BN in caffe version as it is shown in the model zoo for ResNet. Can someone confirm ?
